### DISCUSSION TITLE
How to migrate my WPF application to Linux based application through UNO Platform?

### DISCUSSION THREAD

**Discussion by karasword55** [unoplatform/uno#13312] [General]
Hello, I want to port my existing WPF application to the Uno Platform for the Linux operating system. What are the step-by-step processes I need to follow for this? Which APIs are incompatible, and how can I resolve these issues? Which parts of the code will be different, and how can I address them? Additionally, how can I compare its performance aspect to others like Avalonia, MAUI, and WinUI? I would appreciate your assistance.

**Comment by CalebSerafin** [comment]
Hello 👋, for step-by-step we can find some resources on <https://platform.uno/wpf/> that lay out the steps required.
* There is a written document that outlines the overhaul changes: <https://platform.uno/docs/articles/wpf-migration.html>
* There is a video that first outlines the concepts and then does migration on a code example <https://www.youtube.com/watch?v=doomI2bEBno>.

**Reply by karasword55** [reply]
Thanks a lot. Does the video content provide migrate to linux?

**Reply by CalebSerafin** [reply]
Migrating to UNO is not platform specific. UNO Provides a number of "Heads". Each head compiles to a different UI stack and platform. So if its working with the libraries in UNO, to should mostly work on all the "Heads" the template or command line commands generate.

Some features may be missing from other platforms. For example advanced text decoration and rich text boxes. But it should be stable and the most common functionality should work.

**Reply by CalebSerafin** [reply]
Most of the work is getting the presentation layer fine-tuned (For Linux example). Since after the backend is ported to use .NET 5+, it is cross-platform and can run on 5 Billion Devices™.

**Reply by CalebSerafin** [reply]
So if you make any PInvokes for Windows Libraries, those may not exist on Linux.
Most of .NET Framework exists in .NET 5+. However, windows specific things may need to change.
The way the migration works. Is to create a new Uno project. This is a template that will allow you to select which heads you want. In this case you want either Linux Skia or Linux GTK. The difference is the presentation and rendering library use to display content.
You will then copy and past the code files and XAML files into the base shared project. I reccommend do this incrementally starting with the basic home views and fixing some errors as they arrise (Errors that exist in compile time, Visual Studio's intellisence complains about things that don't exist)

**Reply by sasakrsmanovic** [reply]
Great answers @CalebSerafin ! Thanks for participating and helping out.
